Add reauth flow to ring integration

sdb9696 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add reauthorisation flow to the ring integration.  Currently token expiry requires users to delete and re-create the ring integration which is undesirable.

In order for this to work the ring_doorbell library that this integration uses has some new error types that need to be caught and handled by the integration.  Hence it depended on PR [Bump ring_doorbell to 0.8.0 and handle new exceptions](https://github.com/home-assistant/core/pull/103904) which has now been merged.

N.B. I based these changes heavily on PR https://github.com/home-assistant/core/pull/103351 so many thanks to @dknowles2 and the reviewers on that PR.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

sdb9696 committed:
Add reauth flow to ring integration

allenporter commented on tests/components/ring/test_config_flow.py:
```diff
@@ -40,20 +48,162 @@ async def test_form(hass: HomeAssistant) -> None:
+        "foo@bar.com", "other_fake_password", None
+    )
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": errors_msg}
```
Can you also test where the user fixes the error and goes on to succeed? Helpful just to make sure the combination of flows still gets the user to the right spot.

sdb9696 replied:
Done

allenporter commented on tests/components/ring/test_config_flow.py:
```diff
@@ -40,20 +48,162 @@ async def test_form(hass: HomeAssistant) -> None:
     with patch(
-        "homeassistant.components.ring.config_flow.Auth.fetch_token",
-        side_effect=InvalidAuth,
+        "homeassistant.components.ring.config_flow.ring_doorbell.Auth.fetch_token",
```
While you're here changing this, it could switch to `mock_ring_auth` too

sdb9696 replied:
That's done.  Also done in `test_form` for consistency

allenporter commented on homeassistant/components/ring/manifest.json:
```diff
@@ -13,5 +13,5 @@
   "iot_class": "cloud_polling",
   "loggers": ["ring_doorbell"],
-  "requirements": ["ring-doorbell==0.7.3"]
+  "requirements": ["ring-doorbell==0.8.0"]
```
Can you send this as a separate dependency bump? Makes it easier to review when the PR is doing fewer things.

sdb9696 replied:
In this instance the dependency bump needs to go in at the same time as the code change.  This is because the new exception types in the ring_doorbell library are raised in the case of authentication errors and the oauthlib exceptions that the ring integration currently catches are no longer raised.  If the bump goes in without this change the existing authorisation flow in async_step_user will break.

allenporter replied:
My assumption is that doing the bump and swapping exceptions is a little more mechanical (and can still preserve existing behavior) then actual reauth support is added second. I may have misunderstood though.

sdb9696 replied:
Understood, that's now done, see https://github.com/home-assistant/core/pull/103904.

allenporter commented on homeassistant/components/ring/config_flow.py:
```diff
@@ -56,32 +70,87 @@ async def async_step_user(self, user_input=None):
         )
 
-    async def async_step_2fa(self, user_input=None):
+    async def async_step_2fa_reauth(self, user_input=None):
```
It seems like the 2fa_user and 2fa_reauth flows could be combined, where differences are fairly minimal (just the schema?) and be handled by checking for the presence of `self.reauth_entry`.

I would suggest just inlining `_show_reauth_form` and `_show_user_form` -- while the overall method would be more verbose, its easier to understand what is happening if its in a single place without jumping up and down in the file.

sdb9696 replied:
Both these suggestions are now done.

allenporter reviewed (changes requested):
Overall looks great! I have some suggestions about readability improvements and think its best to do the dependency bump first in a separate PR fixing the compatibility issues, then doing reauth in a follow up. (I'll happily review quickly if you reference the bump pr)

sdb9696 committed:
Refactor re-auth flow post review

sdb9696 commented on homeassistant/components/ring/config_flow.py:
```diff
@@ -56,32 +70,87 @@ async def async_step_user(self, user_input=None):
+            user_input[CONF_USERNAME] = self.reauth_entry.data[CONF_USERNAME]
+            try:
+                token = await validate_input(self.hass, user_input)
+                data = {
```
That's done.  Also done in `async_step_user` for consistency.

sdb9696 commented:
> Overall looks great! I have some suggestions about readability improvements and think its best to do the dependency bump first in a separate PR fixing the compatibility issues, then doing reauth in a follow up. (I'll happily review quickly if you reference the bump pr)

Thanks @allenporter, I have implemented all of your suggestions except for the dependency bump separation due to the reasons given on the other comment and in the PR description.  Hopefully it makes sense and I cannot see another way on this occasion without making things very complicated.  This is a one off issue as going forwards consumers of ring_doorbell will only catch the library specific exceptions rather than catching lower level implementation exceptions.

qq, should I Resolve Conversation myself when I've implemented the suggestion or wait for you to do it?

Many thanks!

allenporter commented on tests/components/ring/test_init.py:
```diff
@@ -32,3 +38,116 @@ async def test_setup(hass: HomeAssistant, requests_mock: requests_mock.Mocker) -
+        assert any(mock_config_entry.async_get_active_flows(hass, {SOURCE_REAUTH}))
+
+
+error_testdata = [
```
I'd say inline this

sdb9696 replied:
Done

allenporter commented on tests/components/ring/test_init.py:
```diff
@@ -32,3 +38,116 @@ async def test_setup(hass: HomeAssistant, requests_mock: requests_mock.Mocker) -
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    if update_type == "global":
```
Here and above: The preferred approach is just have the patch method as the parameterized method, avoiding conditionals in tests. (Then If the behavior varies too much, then just copying the test and code is preferred)

sdb9696 replied:
Fixed

allenporter commented:
> qq, should I Resolve Conversation myself when I've implemented the suggestion or wait for you to do it?

My personal approach is:
- Resolve comments I addressed that I believe were straight forward or non-controversial (e.g. i'm pretty confident I understood and addressed what they were asking me)
- Leave open comments where either there was room for interpretation on what they were looking for and I'm not quite sure if i addressed it, or had additional questions, or I disagreed with the reviewer and have a different suggestion/stance.

sdb9696 committed:
Fix threading issue on device update

MartinHjelmare commented on homeassistant/components/ring/config_flow.py:
```diff
@@ -39,48 +47,95 @@ class RingConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+            else:
+                data = {
+                    CONF_USERNAME: user_input[CONF_USERNAME],
+                    "token": token,
```
It's good to use a constant for strings that are used more than once. I'd use a constant `CONF_TOKEN` for `"token"`.

sdb9696 replied:
Hi Martin, yes agreed. I started to introduce constants with this PR and thought about replacing all the existing instances of "username", "password" and "token" in the integration then decided to do that in a separate PR.  I should have brought the CONF_TOKEN in at this stage but will definitely get to it. (I'm hoping to get the integration to platinum level soon and I know this is a base requirement :)

MartinHjelmare replied:
Good, thanks!
